src/Features/CSharp/Portable/CodeRefactorings/SyncNamespace/CSharpSyncNamespaceService.cs

```diff
@@ -0,0 +1,297 @@
+
+                    var last = members.Last();
+                    var lastWithTrivia = last.WithAppendedTrailingTrivia(triviaFromNamespaceDecl.closingTrivia);
+                    members = members.Replace(last, lastWithTrivia);
+                }
+                else
+                {
+                    eofToken = eofToken.WithPrependedLeadingTrivia(
+                        triviaFromNamespaceDecl.openingTrivia.Concat(triviaFromNamespaceDecl.closingTrivia));
+                }
+
+                return compilationUnit.Update(
+                    compilationUnit.Externs.AddRange(namespaceDeclaration.Externs), 
+                    compilationUnit.Usings.AddRange(namespaceDeclaration.Usings), 
+                    compilationUnit.AttributeLists, 
+                    members,
+                    eofToken).WithAdditionalAnnotations(Formatter.Annotation);
+            }
+            // Change namespace name
+            else
+            {
+                return root.ReplaceNode(namespaceDeclaration, 
+                    namespaceDeclaration.WithName(
+                        CreateNameSyntax(targetNamespaceParts, aliasQualifier: null, targetNamespaceParts.Length - 1)
+                        .WithTriviaFrom(namespaceDeclaration.Name)
+                        .WithAdditionalAnnotations(WarningAnnotation)));
+            }
+        }
+
+        private static SyntaxAnnotation WarningAnnotation 
+            => CodeActions.WarningAnnotation.Create(
+                FeaturesResources.Warning_colon_changing_namespace_may_produce_invalid_code_and_change_code_meaning);
+
+        private static NameSyntax GetQualifiedNameSyntax(NameSyntax node)
+            => node.Parent is QualifiedNameSyntax qualifiedName && qualifiedName.Right == node ? qualifiedName : node;
+
+        private static string GetAliasQualifierOpt(NameSyntax name)
+        {
+            while (true)
+            {
+                switch (name.Kind())
+                {
+                    case SyntaxKind.QualifiedName:
+                        name = ((QualifiedNameSyntax)name).Left;
+                        continue;
+                    case SyntaxKind.AliasQualifiedName:
+                        return ((AliasQualifiedNameSyntax)name).Alias.Identifier.ValueText;
+                }
+                return null;
+            }
+        }
+
+        private NameSyntax CreateNameSyntax(ImmutableArray<string> namespaceParts, string aliasQualifier, int index)
+        {
+            var part = namespaceParts[index].EscapeIdentifier();
+            var namePiece = SyntaxFactory.IdentifierName(part);
+
+            if (index == 0)
+            {
+                return aliasQualifier == null ? (NameSyntax)namePiece : SyntaxFactory.AliasQualifiedName(aliasQualifier, namePiece);
+            }
+            else
+            {
+                return SyntaxFactory.QualifiedName(CreateNameSyntax(namespaceParts, aliasQualifier, index - 1), namePiece);
+            }
+        }
```

Comment:
this feels like this can be written at the shared layer.

Reply:
the part involves alias complicate things a bit. I will give it a try.